ecosystem/op-deployer.md

```diff
@@ -0,0 +1,163 @@
+
+Introduce `op-deployer`, an end-to-end chain deployment tool, and get buy-in on its architecture and implementation.
+
+## Goals
+
+- Provide a canonical way to deploy chains.
+- Enable deploying multiple L2s to a single L1 as part of a single logical Superchain.
+- Make it easier for teams like Protocol and Proofs to maintain their deployment tooling.
+
+## Non-Goals
+
+- Replacing `op-e2e`. This will be discussed later, once `op-deployer` is live.
+
+# Problem Statement + Context
+
+As we've
+discussed [previously](https://docs.google.com/document/d/13f8UoO9j05PJdvAZDWuVJWD1UAIEAfb8VbphBnPE3cA/edit#heading=h.jxalqw91xhd0),
+deploying and upgrading OP Chains remains a complicated and frustrating process. Users are directed to a tutorial
+with [18 manual steps](https://docs.optimism.io/builders/chain-operators/tutorials/create-l2-rollup) to follow in order
+to create their own testnet. Within Labs, we use a combination of different tools depending on where the chain is to be
+hosted:
+
+- **opc** deploys new chains to our cloud infrastructure.
+- The **compose devnet** uses a bunch of custom Python tooling to deploy a chain locally using `docker-compose`.
+- The **Kurtosis devnet** uses a combination of custom deployer scripts and the fault proof's `getting-started.sh` to
+  deploy a chain to a new enclave.
+- The Interop team now has [their own](https://github.com/ethereum-optimism/optimism/pull/11590) devnet setup to deploy
+  a set of interoperable chains locally.
+
+Notably, with the exception of the Interop devnet none of the solutions above enable deploying multiple L2s to a single
+L1 as part of a single logical Superchain. This is a huge gap in functionality, that will only become more magnified as
+time goes on.
+
+To resolve this problem we need set of well-integrated, modular tools that define the canonical way of deploying a
+chain. Users can then leverage these tools to build their own deployment tooling on top.
+
+Much of what's described below is already implemented as part of `opc`. By extracting these concepts out of `opc`
+and into the monorepo, we can make them more accessible to other teams and users.
+
+# Proposed Solution
+
+The high-level architecture is described
+in [Modular Deployments](https://docs.google.com/document/d/13f8UoO9j05PJdvAZDWuVJWD1UAIEAfb8VbphBnPE3cA/edit#heading=h.jxalqw91xhd0),
+so read that document first if you haven't already. This design doc will focus on the implementation details of the
+architecture it describes.
+
+## Deployment Pipeline
+
+The core of the modular deployment tooling is the concept of a _deployment pipeline_. The pipeline consists of stages
+that each perform a single piece of the deployment. The stages each take an input file, and enrich it with the data
+pertaining to that particular stage of the deployment. This allows the output of upstream stages to be cached.
+
+The diagram below outlines what these stages are:
+
+![](./op-deployer/pipeline.png)
+
+## Deployment Intents
```

Comment:
Some questions:
1. Will a deployment intent have a standard interface of some kind?
2. Will a deployment intent be executable from op-e2e?
3. Will a deployment intent be coupled to a strict pre-state? In particular, is an intent to deploy X to a local L1 genesis (aka statedump) different from an intent to produce a list of template deployment txs (broadcast output)? 
4. Will every intent run with the forge tooling? Or can there be intents that are not forge?
5. How does an "interrupted deployment" get recovered? One of the pain-points product hit is that they run forge, then have to CTRL+C because of gas spike issues, and then have to start over from the beginning.
6. Will we integrate the source-code verify step? (API calls to etherscan/blockscout). Ideally I think the deployment-tools produce a dump, such that the verification can happen later, in case a new explorer comes online.
7. Does this tool include `L2Genesis`?
8. Will intents have defaults? Maybe defaults sourced from the superchain-registry?
9. How does a pipeline execute, when it involves intermittent pauses for e.g. deployment-tx signing? Does it persist, and continue from where it left of, after a signing phase maybe?
10. Can we a add a `--noninteractive` kind of flag, that auto-signs with some hotkey, for testing purposes?

Personally I'd also suggest running op-deployer intents fully in Go, or at least having some flexibility to do so.
Otherwise each intent will rely on sub-process calls to forge, intermediate config files and output files, and many extra consistency checks.
With forge-scripts in Go, it can be packaged up, and do exactly as the intent is programmed to, without interacting with anything outside of the process, other than an artifacts tarball. I think this makes it much more reproducible in different environments.
The drawback however is that we don't use standard forge. Maybe with the right abstraction it's interchangeable, and forge is version-checked, inputs are all in unique temporary directories, etc.

Reply:
Answers below:

1. Yes - my strawman interface is the one in the `[intent]` stanza in the design doc.
2. Yes - since each deployment stage is a Go function, it'll be easy to call from `op-e2e`.
3. The idea is that it won't matter where you deploy the chain to. In practice we'll have to implement different "backends" in the deployment tool to support both statedump and deployed L1s.
4. No, intents can run without Forge. I'd actually like to move away from shelling out to Forge whenever possible.
5. Stages commit to the intents TOML to store their state periodically. This allows for the pipeline to recover halfway through.
6. Not right now, but we can add tooling to do this in the future or produce a dump like you said.
7. I'd like it to leverage the Go forge-scripts tooling you built to obviate the need for the `L2Genesis` Forge script.
8. Yes, intents will have defaults. The default config is the standard config. This can be defined in the registry.
9. It will persist its state inside the intents TOML to allow for this kind of behavior.
10. Sure!

In general I agree that we want to run the intents fully in Go. Some stuff requires shelling out to Forge right now, but in the future I'd like to refactor everything we can away from that.